File: LeetCode/leet_04/median_of_two_sorted_arrays.java

// space complexity： O(1)， 因为是尾递归
// time complexity: O(logk)，因为每次k都减少一半，而k=(m+n)/2,所以每次
class Solution {
public double findMedianSortedArrays(int[] nums1, int[] nums2) {
    int length1 = nums1.length;
    int length2 = nums2.length;

    // 如果length1+length2是奇数，也就是一共有奇数个数，那么median左右两边的个数相等，那么left=right；
    // 如果是偶数，那么median是中间两个数的加权和，left 为左边的数的index，right为右边数的index
    int left= (length1+length2 +1)/2;
    int right = (length1 + length2 + 2)/2;

    return (getKth(nums1,0,length1-1,nums2,0,length2-1,left) + getKth(nums1,0,length1-1,0,length2-1,right)) /2;
}

    private int getKth(int[] nums1, int start1, int end1, int[] nums2, int start2, int end2, int k) {
        // 两个子数组的元素个数
        int len1 = end1 - start1  + 1;
        int len2 = end2 - start2 + 1;

        // 如果有子数组为空，那么一定是nums1为空
        if (len1 > len2) return getKth(nums2,start2,end2,nums1,start1,end1,k);

        // If left half is empty, then we just need to return kth number of number2
        if (len1 == 0) return nums2[start2 + k -1];

        // If we need to find the 1th number, we just need to return the smaller one of two arrays
        if (k == 1) return math.Min(nums1[start1],nums1[start2]);

        // In case that the number of array is less than k/2, if so, we just drop all the element on the left
        int i = start1 + math.Min(len1,k/2) -1;
        int j = start2 + math.Min(len2, k/2) -1;

        // divide, drop half of k element
        if (nums1[i] > nums2[j]) {
            return getKth(nums1,start1,end1,nums2,j+1,end2, k-(j - start2) +1);
        } else {
        return getKth(nums1,i+1, end1, nums2, start2,end2,k - (i - start1) + 1);
        }
    }
}

File: LeetCode/leet_04/median_of_two_sorted_arrays.go
package leet_04

import "math"

/*
 *	Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.
 *  Follow up: The overall run time complexity should be O(log (m+n)).
 */

/*
	题解见同文件夹markdown。
 */

func getMax(x,y int) int {
	if x > y {
		return x
	}

	return y
}

func getMin(x,y int) int {
	if x > y {
		return y
	}

	return x
}

// time complexity: O(log(min(m,n))
// detail see markdown
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	if len(nums1) > len(nums2) {
		return findMedianSortedArrays(nums2,nums1)
	}

	x,y := len(nums1),len(nums2)
	low,high := 0,x
	var maxLeftX,maxLeftY,minRightX,minRightY int
	for low <= high {
		partitionX := (low+high)/2
		partitionY := (x+y+1)/2 - partitionX

		// edge condition
		if partitionX == 0 {
			maxLeftX = math.MinInt32
		} else {
			maxLeftX = nums1[partitionX-1]
		}
		if partitionX == x {
			minRightX = math.MaxInt32
		} else {
			minRightX = nums1[partitionX]
		}

		if partitionY == 0 {
			maxLeftY = math.MinInt32
		} else {
			maxLeftY = nums2[partitionY-1]
		}
		if partitionY == y {
			minRightY = math.MaxInt32
		} else {
			minRightY = nums2[partitionY]
		}

		// We found the correct place
		// If the total length is odd number, then we choose the  bigger one between maxLeftX and maxLeftY
		// If the total length is even number, then we choose the the average between the max of maxLeft* and minRight*
		if maxLeftX <= minRightY && maxLeftY <= minRightX {
			if (x+y)%2 == 0{
				return (float64(getMax(maxLeftY,maxLeftX)) + float64(getMin(minRightY,minRightX)))/2
			} else {
				return float64(getMax(maxLeftX,maxLeftY))
			}
		} else if maxLeftX > minRightY {
			// We go far right, we need move left
			high = partitionX -1
		} else  {
			// We go far left, we need move right
			low = partitionX + 1
		}
	}
	return 0.0
}


